homeassistant/components/geo_location/nsw_rural_fire_service_feed.py

```diff
@@ -0,0 +1,244 @@
+
+    def _update(self):
+        """Update the feed and then update connected entities."""
+        import geojson_client
+        status, feed_entries = self._feed.update()
+        if status == geojson_client.UPDATE_OK:
+            _LOGGER.debug("Data retrieved %s", feed_entries)
+            # Keep a copy of all feed entries for future lookups by entities.
+            self._feed_entries = feed_entries.copy()
+            keep_entries = self._update_or_remove_entities(feed_entries)
+            self._generate_new_entities(keep_entries)
+        elif status == geojson_client.UPDATE_OK_NO_DATA:
+            _LOGGER.debug("Update successful, but no data received from %s",
+                          self._feed)
+        else:
+            _LOGGER.warning("Update not successful, no data received from %s",
+                            self._feed)
+            # Remove all entities.
+            self._update_or_remove_entities([])
+
+    def _update_or_remove_entities(self, feed_entries):
+        """Update existing entries and remove obsolete entities."""
+        _LOGGER.debug("Entries for updating: %s", feed_entries)
+        # Remove obsolete entities for events that have disappeared
+        managed_entities = self._managed_entities.copy()
+        entries_lookup = {entry.external_id: entry for entry in feed_entries}
+        for entity in managed_entities:
+            entry = entries_lookup.pop(entity.external_id, None)
+            if entry is None:
+                # Remove obsolete entity.
+                _LOGGER.debug("Entity not current anymore %s", entity)
+                self._managed_entities.remove(entity)
+                self._hass.add_job(entity.async_remove())
+                continue
+            # Existing entity - update details.
+            _LOGGER.debug("Existing entity found %s", entity)
+            feed_entries.remove(entry)
+            entity.schedule_update_ha_state(True)
+        # Return the remaining entries that new entities must be created for.
+        return feed_entries
+
+    def _generate_new_entities(self, entries):
+        """Generate new entities for events."""
+        new_entities = []
+        for entry in entries:
+            new_entity = NswRuralFireServiceLocationEvent(self, entry)
+            _LOGGER.debug("New entity added %s", new_entity)
+            new_entities.append(new_entity)
+        # Add new entities to HA and keep track of them in this manager.
+        self._add_entities(new_entities, True)
+        self._managed_entities.extend(new_entities)
+
+    def get_feed_entry(self, external_id):
+        """Return a feed entry identified by external id."""
+        return next((entry for entry in self._feed_entries
+                     if entry.external_id == external_id), None)
+
+
+class NswRuralFireServiceLocationEvent(GeoLocationEvent):
+    """This represents an external event with GeoJSON data."""
+
+    def __init__(self, feed_manager, feed_entry):
```

Comment:
Pass in the `external_id` instead of the `feed_entry` and store the `external_id` as an instance attribute in init. We will update and get the feed entry after entity addition anyway.

Reply:
OK - only passing in external id now